state.go

```diff
@@ -478,115 +533,184 @@ func (a *AuthorizeState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if len(installers) < 1 {
+		return uc.HandleError(ctx, c, NewTransientError(
+			errors.New("GetInstallers() returned empty list? Should not happen")))
+	}
+	err = installers[0].CommitUpdate()
+	if err != nil {
+		// we need to perform roll-back here; one scenario is when
+		// u-boot fw utils won't work after update; at this point
+		// without rolling-back it won't be possible to perform new
+		// update
+		merr := NewTransientError(errors.Errorf("update commit failed: %s", err.Error()))
+		return uc.HandleError(ctx, c, merr)
 	}
 
-	// HasUpgrade() returned false
-	// most probably booting new image failed and u-boot rolled back to
-	// previous image
-	log.Errorf("update info for deployment %v present, but update flag is not set;"+
-		" running rollback image (previous active partition)",
-		uv.Update().ID)
+	// After this it is too late to roll back, so indidate that DB schema
+	// migration is now permanent, if there was one.
+	uc.Update().HasDBSchemaUpdate = false
 
-	return NewRollbackState(uv.Update(), false, false), false
+	// And then store the data together with the new artifact name,
+	// indicating that we have now migrated to a new artifact!
+	err = StoreStateDataAndTransaction(ctx.store, datastore.StateData{
+		Name:       uc.Id(),
+		UpdateInfo: *uc.Update(),
+	}, func(txn store.Transaction) error {
+		log.Debugf("Committing new artifact name: %s", uc.Update().ArtifactName())
+		return txn.WriteAll(datastore.ArtifactNameKey, []byte(uc.Update().ArtifactName()))
+	})
+	if err != nil {
+		log.Error("Could not write state data to persistent storage: ", err.Error())
+		return handleStateDataError(ctx, NewUpdateErrorState(NewTransientError(err), uc.Update()),
+			false, datastore.MenderStateUpdateInstall, uc.Update(), err)
+	}
+
+	// Do rest of update commits now; then post commit-tasks
+	return NewUpdateAfterFirstCommitState(uc.Update()), false
 }
 
-type UpdateCommitState struct {
-	UpdateState
+type UpdatePreCommitStatusReportRetryState struct {
+	waitState
+	returnToState State
+	reportTries   int
 }
 
-func NewUpdateCommitState(update client.UpdateResponse) State {
-	return &UpdateCommitState{
-		UpdateState: NewUpdateState(MenderStateUpdateCommit,
-			ToArtifactCommit, update),
+func NewUpdatePreCommitStatusReportRetryState(returnToState State, reportTries int) State {
+	return &UpdatePreCommitStatusReportRetryState{
+		waitState: waitState{
+			baseState: baseState{
+				id: datastore.MenderStateUpdatePreCommitStatusReportRetry,
+				t:  ToArtifactCommit_Enter,
+			},
+		},
+		returnToState: returnToState,
+		reportTries:   reportTries,
 	}
 }
 
-func (uc *UpdateCommitState) Handle(ctx *StateContext, c Controller) (State, bool) {
+func (usr *UpdatePreCommitStatusReportRetryState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	maxTrySending :=
+		maxSendingAttempts(c.GetUpdatePollInterval(),
+			c.GetRetryPollInterval(), minReportSendRetries)
+		// we are always initializing with triesSending = 1
+	maxTrySending++
```

Comment:
Please align comment.

It took me a while to figure out what it meant. I think a cleaner alternative could be to have this +1 increment already accounted for in `maxSendingAttempts` logic.

I am not certain about the implication of such change, so better not to change it now....

Reply:
Yes, I tend to agree. Enough change already! :-)